crates/xtask/src/changelog.rs

```diff
@@ -0,0 +1,549 @@
+    ///
+    /// Validates requirements in docs/contributing/changelog.md
+    fn parse(contents: String) -> Result<Changelog> {
+        let mut skip_counter = 0;
+        let mut releases: Vec<Release> = Vec::new();
+
+        let mut lines = contents.lines().filter(|line| !line.is_empty()).peekable();
+
+        // First line is H1.
+        if lines.next() != Some("# Changelog") {
+            return Err(anyhow!("H1 with 'Changelog' is required"));
+        }
+
+        while let Some(mut current_line) = lines.next() {
+            let mut next_line = lines.peek();
+
+            // Parse a whole release
+            if let Some(version_string) = current_line.strip_prefix("## ") {
+                let mut release = Release::from(version_string)?;
+
+                // Parse each release type (major, minor, patch)
+                while next_line.is_some_and(|line| line.starts_with("### ")) {
+                    // Advance
+                    current_line = lines.next().unwrap();
+                    next_line = lines.peek();
+
+                    let release_type_string = current_line.trim_start_matches("### ");
+                    let release_type = ReleaseType::from_str(release_type_string, true)
+                        .map_err(|err| anyhow!("Failed to parse version_type: {err}"))?;
+
+                    let mut release_descriptions = Vec::new();
+
+                    // Parse the release descriptions
+                    // Some lines may be a new description. Some lines may be continuation of
+                    // previous descriptions.
+                    while next_line
+                        .is_some_and(|line| line.starts_with("- ") || line.starts_with("  "))
+                    {
+                        // Advance
+                        current_line = lines.next().unwrap();
+                        next_line = lines.peek();
+
+                        // New description
+                        if current_line.starts_with("- ") {
+                            release_descriptions
+                                .push(current_line.trim_start_matches("- ").to_string());
+                        } else if current_line.starts_with("  ") {
+                            // Append to previous description
+                            let mut last = release_descriptions
+                                .pop()
+                                .expect("No last description to append to.");
+                            last.push_str(format!("\n{current_line}").as_str());
+                            release_descriptions.push(last);
+                        }
+                    }
+
+                    match release_type {
+                        ReleaseType::Major => release.major = release_descriptions,
+                        ReleaseType::Minor => release.minor = release_descriptions,
+                        ReleaseType::Patch => release.patch = release_descriptions,
```

Comment:
Actually, I think it might be beneficial to use a map for the descriptions: `BTreeMap<ReleaseType, Vec<String>>` with the invariant that values are not empty and there is at least one key.